### PR TITLE
Try: Refactor contextual toolbar to work better with floats

### LINKED ISSUE #4764: Block control for adjacent blocks gets messed up when one of them is left/right aligned
## Issue Overview

Create two blocks one after the other, and left or right align any one of them. The other block, which is center aligned, has its block control extend over the entire area occupied by the former block. Also, the overflow More Options menu does not appear for the center aligned block. Thus, trying to delete from the More Options menu ends up deleting the left/right aligned block. 

## Steps to Reproduce (for bugs)

1. Create any two blocks that can have an align option, one below the other.
2. Left or right align either one of them
3. Click the center aligned block. Observe the block border and block control extend over the area occupied by the other block.
4. Try to delete the center aligned block. Hover mouse to the More options menu, and click delete. The center aligned block will not get deleted, but the other block gets deleted.
5. This issue can be extended to multiple blocks. Create several blocks, align all except one to the left or right. Then click the center aligned block to observe the boundary for the block control.

## Expected Behavior

1. The boundary should enclose only the selected block, and not cover the other blocks.

## Screenshots / Video

The gif below shows multiple adjacent blocks, and clicking the center aligned block has its  block control overlay its preceding blocks.

d1

The gif below shows how it's impossible to delete the center aligned block (or perform any other operation from the More Options menu), since the menu option is activated only for the smaller, right aligned block.

d2

### PR DESCRIPTION
**Pull request by jasmussen**
This PR seeks to fix #4764.

But it is also a rather expansive refactor, and big surgery, this late in the phase. I believe it is a good refactor and relatively solid. But because it has the potential to cause breakage in many aspects, like editor styles and exotic combinations of blocks, RTL, probably other aspects too, not only will this need a lot of testing, but there may be enough dragons that it could be worth punting this to phase 2.

This PR does a number of things:

- It moves the contextual toolbar into the block edit div, which is the one we float. This helps make it _connected_ to the content.
- Add some complex clearing rules so we avoid many of the gnarly situations where a selected block after a floated block has a weirdly tall size.
- Fixes so a paragraph block that follows a float does behave as it will on the frontend (i.e. won't clear), but also has a toolbar that is correctly positioned.

This moving around of things caused subsequent issues, which means this PR also:

- Fixes the toolbar appearance on mobile.
- Improves upon the appearance of the toolbar on floated items on mobile.
- Fixes hover label positioning, not only so they work with floats, but are positioned correctly as a result of this refactor.
- Fixes issues with wide and fullwide toolbar positioning.

**Screenshots of what it fixes**. The hover label is now correctly positioned on floats:

screenshot 2018-11-01 at 11 02 33

The toolbar of the 2nd floated block is now correctly positioned:

screenshot 2018-11-01 at 10 54 51

The boundaries and toolbar of a 3rd block that follows 2 floated blocks, is now correct:

screenshot 2018-11-01 at 10 54 55

Text following one or more floated blocks will now have correct boundaries and toolbar:

screenshot 2018-11-01 at 12 02 51

However there is a new issue with multi-selection. It may be hard to fix for technical reasons, but it means the multi select indicator for floats is currently not super visible. This branch:

screenshot 2018-11-01 at 12 12 43

Master:

screenshot 2018-11-01 at 12 12 28

However we can do something else for multiselected floats. But before I do that, I'd like feedback on this PR first. Please give it a thorough shake.

### CODE REVIEW

**Comment by GlennMartin1** [comment]
Didn't test it, but it looks promising. 

> but there may be enough dragons that it could be worth punting this to phase 2.

Bring the dragons now before 5.0.

**Review by alexislloyd — COMMENTED** [review]
Design functionality looks good here. I think there are some enhancements around floats that we can tackle in the future to make the overall experience better, but that's a bigger issue. I can't speak to the difficulty of thoroughly testing this in re: considerations for 5.0 release vs. pushing to 5.1.

**Comment by designsimply** [comment]
I tested `try/float-fixes` (56d8735bd) with the following steps using Firefox 63.0 on macOS 10.13.6 and found that the selection appeared correctly for a centered instagram block below a right-aligned gallery block. ([21s](https://cloudup.com/cYxrzj-15si))

@jasmussen I'm not sure if this is a thorough enough test, but I did find that it solves the test case noted in #4764.

**Comment by ZebulanStanphill** [comment]
I found an issue.

`master`:
image

This PR:
image

The next block after a floated block is pushed below the floated block, rather than appearing below the previous non-floated block.

The front-end still works properly, as you would expect:
image

Aside from that unfortunate issue, however, this works great! This fixes #7491 and makes the UI around floated blocks feel _so_ much nicer!

I wonder... could you bring back the movement controls with this change? Along with fixing the aforementioned issue, that would practically perfect the floated block UI!

**Comment by jasmussen** [comment]
Thanks for that report, Zeb, will take a look. 

I will also add some comments to the code, for specific parts that I'm slightly nervous about and want code sanity checks on. 

>  I'm not sure if this is a thorough enough test, but I did find that it solves the test case noted in #4764.

Really appreciate your testing, so thorough, much appreciated. Things that I would want us to make sure are solid:

- Sequences of multiple floats after each other, resized or not, left or right
- Wide, fullwide images or blocks, before and after floats
- Floats for every floatable block type
- Text between two floats, or text after floats
- Every other block type, following a float. 
- The above, on mobile, medium and large breakpoints. 
- The above, with unified toolbar

File: packages/edit-post/src/components/visual-editor/style.scss [hunk 1]
```diff
@@ -84,6 +86,12 @@
 }
 
 .edit-post-visual-editor {
+	// If the first block is floated, it needs top margin, unlike the rule in line 69.
```
Comment (jasmussen) [inline packages/edit-post/src/components/visual-editor/style.scss]: This is a little awkward to have a special rule like this. But it is due to the negative margins that every block, except floats, have. It would be nice to refactor this rule to be unnecessary in the future. Whether this is through removing all the negative margins by making them unnecessary (unlikely) or whether it's through changing even floats so they behave the same as other blocks. 

If the title was an actual block, then it would be a non issue.

File: packages/editor/src/components/block-list/style.scss [hunk 2]
```diff
@@ -820,39 +845,50 @@
...
-	@include break-small() {
+		// This prevents floats from messing up the position.
 		position: absolute;
 		left: 0;
 	}
+}
 
-	@include break-small() {
-		width: auto;
+// This rule ensures that any blocks that are not a Paragraph, that follows any aligned block, clears them.
```
Comment (jasmussen) [inline packages/editor/src/components/block-list/style.scss]: I don't like that we are creating special rules on a per-block basis. 

In this case, it was the only way I could think of, to ensure that the following worked as intended:

screenshot 2018-11-02 at 09 45 57

To elaborate — unless this bottom placeholder block explicitly _clears_ the preceding floats, then the block outline will be painted all the way up to before the first float. This is because the block techically floats after the first two, so the outline starts all the way at the top. Except this is not the user expectation.

On the flipside, we explicitly allow the Paragraph block to not clear the preceding floats, so that this can work:

screenshot 2018-11-02 at 09 47 58

This is the normal floats behavior, and it "looks correct" for the user because the text snakes in between the floats.

**Comment by jasmussen** [comment]
There are legitimate test failures around multi selection. Specifically it seems you can't select two blocks, and click the ellipsis menu. It's keyboard but not mouse accessible.

I'm looking into it, but it appears to be related to the issue we've been seeing with floated images in general, where it can be very easy to invoke a multi selection even just clicking a float. It would be nice if I could solve both issues at the same time, but on the flipside, this bug is also a blocker for this branch, and needs to be fixed for it to be merged. @iseulde do you have time to lend your eyes?

**Comment by jasmussen** [comment]
Okay, so I think I managed to fix the issue [outlined here](https://github.com/WordPress/gutenberg/pull/11357#issuecomment-435324871), but I did so by adding yet another div. Outside of it working for me now, we'll know as well as the tests should come out green when they're finished. 

The relevant code is here: https://github.com/WordPress/gutenberg/pull/11357/commits/312921f4f48526ebf553f6264e86a28d8e663414#diff-8b707a252d7b6da3af477595c2ce11dcR525

As you can see, this is _yet another wrapping div_. To explain why it's there on a high level, it's because if we need the toolbar to be positioned in proximity to the block content, it has to be part of the div that gets floated, or otherwise manipulated. But the element that previously had the `editor-block-list__block-edit` className — [this element](https://github.com/WordPress/gutenberg/pull/11357/commits/312921f4f48526ebf553f6264e86a28d8e663414#diff-8b707a252d7b6da3af477595c2ce11dcR533) — then the toolbar would inherit the dragstart actions that invoke multi select. 

Is this additional div bad? You tell me. I know that additional markup is intrinsic complexity, small though it may be. In this case, though, it I feel it might make sense to be able to group additional elements in visual proximity of the content, especially as we move into phase 2. In other words, it feels worth it to me.

**Comment by ZebulanStanphill** [comment]
@jasmussen I'm actually using my own custom child theme for Divi that ports its Classic Editor TinyMCE styling to Gutenberg. However, I tried switching to Twenty Seventeen and the issue still occurred there, so it definitely isn't a theme styles issue.
image

Here is the markup of the post:
```html
<!-- wp:image {"id":495,"align":"right","width":251,"height":140} -->
<div class="wp-block-image"><figure class="alignright is-resized"><img src="https://zebulan.com/wp-content/uploads/2018/10/blog-looking-back-sgz-plays-minecraft-e1.png" alt="" class="wp-image-495" width="251" height="140"/></figure></div>
<!-- /wp:image -->

<!-- wp:paragraph -->
<p>blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah&nbsp;blah</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah&nbsp;blah</p>
<!-- /wp:paragraph -->
```

I updated to the latest version of the branch, but the issue is still happening.

Also, with regard to the multi-selection issue and wrapping `<div>`s, I think the latter may end up being necessary anyway because of issues like this:
image
The toolbar doesn't remain sticky for the entire multiple block selection because it is anchored to the first block, so scrolling past it causes the toolbar to go off-screen.

**Comment by chrisvanpatten** [comment]
I really want to test this, but I'm AFK (and trying really hard to stay that way!) until Monday. That said, it looks like a good iterative improvement. Keep it up 👍

**Comment by jasmussen** [comment]
Okay, after a good old food poisoning, I managed to give this a good rebase. 

@ZebulanStanphill I still can't reproduce your issues. Tried in 4.9.8 and 5.0 with TwentySeventeen and without. Tried in Chrome, Firefox, Safari, all seeing the same, which is that the 2nd paragraph block clears the image. Here's what I see:

screenshot 2018-11-06 at 11 50 43

screenshot 2018-11-06 at 11 56 14

I even tried adding a wide main editor style:

screenshot 2018-11-06 at 12 00 05

Is there any chance you could try temporarily disabling divi and see if that fixes the issue? 

> The toolbar doesn't remain sticky for the entire multiple block selection because it is anchored to the first block, so scrolling past it causes the toolbar to go off-screen.

This is a separate issue, which I think is tracked separately. There are ideas for how to fix this, but it's not easy, and I don't know that this wrapping div I'm creating (which is on a per block basis) will help. I think we need to on the fly create a wrapping div around a multi selection, which is complex.

**Comment by jasmussen** [comment]
The failing tests are confounding me. I would appreciate any help in fixing them. For all intents they should pass, but they fail for me locally. 

screenshot 2018-11-06 at 13 09 16

**Comment by jasmussen** [comment]
@youknowriad @mtias If any of you have time to look at this branch, code wise, and experience wise, I'd appreciate input on whether you'd like to see this in the 5.0 release or if it should be punted. I'm personally increasingly liking this approach, but I'm also very much aware that it's major surgery.

Here's some good demo content to test:

```
<!-- wp:image {"align":"left"} -->
<div class="wp-block-image"><figure class="alignleft"><img alt=""/></figure></div>
<!-- /wp:image -->

<!-- wp:image {"align":"right"} -->
<div class="wp-block-image"><figure class="alignright"><img alt=""/></figure></div>
<!-- /wp:image -->

<!-- wp:image -->
<figure class="wp-block-image"><img alt=""/></figure>
<!-- /wp:image -->

<!-- wp:paragraph -->
<p>Loem ipsum dolor sit amet, ferri vidisse nam eu, ad nec copiosae mnesarchum vituperatoribus. Te brute dicunt sea, ut vis omnium menandri, ut sumo aliquam has. Eum aperiam interpretaris at, sea et recusabo expetenda, omnis tibique mea no. Pri suas partem ea, ius sonet numquam offendit cu, ad simul admodum pri. Eum cu unum choro albucius.</p>
<!-- /wp:paragraph -->
```

**Comment by ZebulanStanphill** [comment]
@jasmussen I tried with Twenty Seventeen and the issue still happened. I found the CSS causing the issue:

```css
.editor-block-list__block-edit::after {
    display: block;
    content: "";
    clear: both;
}
```
Which is coming from line 19 of `/wp-content/plugins/editor-blocks/dist/blocks.editor.build.css?ver=1.1.0`

Removing either the `content` or the `clear` properties fixes the issue.

Tested in both Firefox Developer Edition 64 and Chromium 70.

**Comment by jasmussen** [comment]
Wait, that line seems to stem from a plugin called `editor-blocks`. Do you by chance have [this plugin](https://wordpress.org/plugins/editor-blocks/) installed, @ZebulanStanphill? If yes, can you try to temporarily disable it and see if that fixes your issue? 

If it does, then that's a clearing rule they're adding to every block, and they probably shouldn't be, and we can submit a bug report to their plugin.

**Comment by jasmussen** [comment]
As a teaser for why I'd appreciate eyes on this, here's how `master` looks at the moment:

master

Here's this branch:

this branch

**Comment by jasmussen** [comment]
Noting that I've tested rather thoroughly with a variety of blocks and the Twenty themes, and I'm starting to think that the value this refactor brings outweighs the theoretical downsides. I'm sure it could use more testing, but it might be worth getting this in soon, even if we have to revert it later. 

Adding to a milestone for that reason.

**Comment by ZebulanStanphill** [comment]
@jasmussen Yep, I had that plugin installed. Thanks for the help. :+1:

File: packages/editor/src/components/block-list/block.js [hunk 3]
```diff
@@ -515,45 +515,46 @@ export class BlockListBlock extends Component {
...
-											block={ block }
-										/>,
-										<div key="invalid-preview">
-											{ getSaveElement( blockType, block.attributes ) }
-										</div>,
-									] }
-								</BlockCrashBoundary>
-								{ shouldShowMobileToolbar && (
-									<BlockMobileToolbar
+							<div className="editor-block-list__block-edit">
```
Comment (youknowriad) [inline packages/editor/src/components/block-list/block.js]: Can you explain a bit why the changes here. These are the most impacting changes I think.

File: packages/editor/src/components/block-list/style.scss [hunk 4]
```diff
@@ -820,39 +845,50 @@
...
+// This is necessary to ensure the selected block outlines and toolbar are correctly positioned.
+.editor-block-list__block[data-align] + .editor-block-list__block:not([data-align="right"]):not([data-type="core/paragraph"]) {
```
Comment (youknowriad) [inline packages/editor/src/components/block-list/style.scss]: Why is the paragraph block special here?

**Comment by youknowriad** [comment]
Some z-index weirdness on mobile

screen shot 2018-11-08 at 09 06 04

File: packages/editor/src/components/block-list/block.js (hunk 3 above)
Comment (youknowriad) [reply packages/editor/src/components/block-list/block.js]: I think I got it from the description of the PR

File: packages/editor/src/components/block-list/style.scss (hunk 4 above)
Comment (youknowriad) [reply packages/editor/src/components/block-list/style.scss]: This is the most concerning part for me but I don't see anything special about the paragraph block. Heading blocks are similar, lists are similar, quotes are similar. We could also image custom text blocks... 

I don't really know what these rules are doing, but we should definitely try to do without any block targetting here.

**Comment by jasmussen** [comment]
Thank you for the review. Responding here so the comments won't get buried in a collapsed "outdated" review.

> This is the most concerning part for me but I don't see anything special about the paragraph block. Heading blocks are similar, lists are similar, quotes are similar. We could also image custom text blocks...

> I don't really know what these rules are doing, but we should definitely try to do without any block targetting here.

Yep. Completely agree. Let me try and explain those rules with a few examples.

**1: blocks that _should_ clear preceding blocks:**

Here's master:

master

Here's this branch:

thisbranch

As you can tell in this example, the image block (could be any block that is _block level_) should clear preceding floats. In master the block itself DOES clear those floats, but as you can see the block outline and the block toolbar _do not_ clear preceding floats. Basically because they "fit", i.e. they snake in between the space left over by the floats. So in order for those borders and toolbars to behave as a user would expect them to, we need to explicitly clear them if a block is _block level_. 

**2: blocks that _should not_ clear preceding blocks:**

Here's master:

master-text

Here's this branch:

thisbranch-text

Again, this is about meeting the user expectation of how the UI should behave. What you're seeing in this branch is the specific rule applied to paragraphs in practice. What it does is make sure that the text content does not clear the floats, but allows the text to flow around them as it does in HTML. It also adds a special rule to the toolbar, to ensure the toolbar doesn't float, this lets it sit far to the left as would be user expectation.

**Is there a better way?**

I'm not sure. The rules are not so much specific to the Paragraph block, as much as they are specific to _block level blocks_ vs. _inline blocks_. 

If we remove those rules, we'll still have some of the benefits, specifically around the positioning of the hover labels and toolbars on floats:

screenshot 2018-11-08 at 09 56 23

screenshot 2018-11-08 at 09 56 31

However we lose the improvements that are made to any non floated blocks that follow floats:

screenshot 2018-11-08 at 09 55 56

screenshot 2018-11-08 at 09 56 18

**Comment by youknowriad** [comment]
@jasmussen My naive thinking is that we shouldn't clear floats at all as this is the expected behavior in frontend for all floated blocks, and editor should behave similarly. What's wrong with not clearing floats?

**Comment by youknowriad** [comment]
I guess what could go wrong is the first screenshot. Can't we make the toolbar 100% wide (like the block content) this way it will clear itself without an explicit rule.

**Comment by youknowriad** [comment]
> I think that's a good challenge. I will give it another stab. But I don't know that we can make the outlines clear as they are supposed to, without the clearing rule. It's in that gray area where user expectations suggest one look, while the technical underpinnings present another.

My thinking is that with "floats", the natural "clearing" happens when the width of the block content is bigger than the available space, so if we want the toolbar to clear similarily to the block content, we should try to make their widths equivalent.

 - The block's wrapper width should be the same as its content (which I think is 100% in most cases except other floated blocks)
 - The toolbar if it's already moved inside this block wrapper can have a 100% width.

(That's just theory though, I know it can be worse in practice :) )

**Comment by jasmussen** [comment]
> My thinking is that with "floats", the natural "clearing" happens when the width of the block content is bigger than the available space, so if we want the toolbar to clear similarily to the block content, we should try to make their widths equivalent.

Yes, this is correct. And by the way you are commenting on exactly what I'm the most worried about in this PR, so it is exactly the kind of sanity check that I asked for. 

And I think a fullwide toolbar like that _can work_ — but the outlines won't. The outlines are painted outside the block, and because a non-clearing block essentially starts at the top of the floats. 

But I have a different idea, I'll get back to you.

**Comment by jasmussen** [comment]
So, here's the idea. We have a single universal rule that applies to _any block that follows a floated block_:

```
// This rule ensures that any block that follows a float, has a better positioned contextual toolbar.
.editor-block-list__block[data-align] + .editor-block-list__block:not([data-align="right"]) {
	.editor-block-contextual-toolbar {
		float: none;
	}
}
```

The result is this:

screenshot 2018-11-08 at 10 57 49

Cool.

But also this:

screenshot 2018-11-08 at 10 57 41

Less cool. But it is _same rule for all_, and it's better than this:



What do you think?

I also tried the idea you suggested here:

> My thinking is that with "floats", the natural "clearing" happens when the width of the block content is bigger than the available space, so if we want the toolbar to clear similarily to the block content, we should try to make their widths equivalent.

But this doesn't really work. As noted, the outline still goes to the top:

screenshot 2018-11-08 at 11 01 54

But text is far worse, because the text doesn't clear the floats, but the block toolbar does:

screenshot 2018-11-08 at 11 02 03

And then we're back to making special exceptions.

**Comment by youknowriad** [comment]
> So, here's the idea. We have a single universal rule that applies to any block that follows a floated block:

Why do we even need to float the toolbar in the first place?

**Comment by jasmussen** [comment]
> Why do we even need to float the toolbar in the first place?

Excellent question, and happy to answer it, because it's relatively important. It's also documented here: https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/block-list/style.scss#L818

In order for the toolbar to be able to behave as a _sticky_ toolbar (that is, it scrolls with you down the page), it has to adhere to different rules than usually abs positioned elements (it behaves more like a relatively positioned element). But as all such elementsm, it has dimensions. Which means a selected block with the block toolbar showing is taller than an unselected block. This led to very complicated CSS back in the day, having to compensate with negative margins when the toolbar was there. 

By floating the toolbar — any float — we can reduce the footprint of the block toolbar by essentially taking it out of the flow. That way we don't have to compensate for its dimensions.

Also, without this float, we cannot collapse block margins, which is important to make it easier to style the editor.

To put it simply, floating the toolbar is necessary for the block UI to not affect the dimensions of the block.

**Comment by jasmussen** [comment]
> Won't this "reset" affect those use cases:

Huh, that is curious. 

Removing that rule entirely, even in master, appears to not do anything. I know for a fact that it was necessary in the past, I spent months on that PR refactoring it to work. 

I'm going to ... step away from the computer for a bit and make sure I haven't lost it. 

Though I'm going to double check after this — I'm going to retrace my steps, I want to make sure that a giant toolbar doesn't break it for example. (Amazing catch, in other words)

However if it turns out, that the float is truly unecessary, we're down to even fewer rules. But we're still left with this:

screenshot 2018-11-08 at 11 34 54

In this screenshot, the 3rd block is selected. It's "technically accurate" because it floats after the two preceding ones. But if we are okay with this tradeoff, it definitely does make this PR more resilient. What do you think?

**Comment by youknowriad** [comment]
I think it's an ok tradeoff for me and about the "float" necessity, last time I saw this, I think it was necessary for mobile or wide-alignments somehow, but maybe we can find a better alternative.

**Comment by jasmussen** [comment]
Okay, fixed a few things and pushed. This still needs good testing due to the removal of the float rule.

**Comment by jasmussen** [comment]
Rebased this, and awaiting the tests to see if they pass. After the changes made as a result of feedback, I would suggest the risk of merging this PR has gone down a great deal. Testing still works out fine in all cases, and although it now no longer goes quite as far as it did initially, the removal of block-type-specific rules makes it a much more generic solution.

I feel it would be good to ship this.

**Comment by gziolo** [comment]
Classic block regresses in this PR:
screen shot 2018-11-09 at 12 56 06

Is this intentional that border is missing for floated block?
screen shot 2018-11-09 at 13 08 38

**Comment by gziolo** [comment]
It's very hard to use Gallery block when it's floated. See:

floated-blocks

**Comment by jasmussen** [comment]
Fixed a small regression with the classic block ellipsis menu. It now looks right again:

screenshot 2018-11-09 at 13 12 31

> Is this intentional that border is missing for the floated block?

Yes. This is the same in master, and goes back to a limitation on how floats work in general. I could see us exploring a refactor in the future, but it would be very complex due to how floats work.

> It's very hard to use Gallery block when it's floated. See:

This looks like a longstanding issue with multi select being slightly too easy to invoke in some sitautions. Can you confirm whether or not this behavior is regressed in _this_ branch vs. master? Last I checked the behavior was the same.

**Review by youknowriad — APPROVED** [review]
Floats are still not great but floats are better.

**Comment by jasmussen** [comment]
Thank you Riad! Merging.

**Comment by bph** [comment]
Testing this via  4.3 rc1 and seems to work very nicely with image or pullquote. I also tried it with an audio block and while I am in the block I see the toolbar but when I click out site the editor, the audioblock disappears from the edit view. It's still there, though as it displays on the front end. 
Here is the video. http://recordit.co/898tjKTCao

**Comment by ktmn** [comment]
Did this PR cause this: 

In an alignfull block, the nested block's toolbar is off and so is the hover label.

Tested on frontenberg, this is the source:

```
<!-- wp:columns {"align":"full"} -->
<div class="wp-block-columns alignfull has-2-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>asd</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>qwe</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
```

I'm also having a really hard time selecting nested blocks now but can't really pinpoint it. Sometimes I hover a block, the blue borders of highlight appear, I left click, nothing happens. I left click some more, nope, dead clicks, won't select the block...

**Comment by jasmussen** [comment]
Possibly, yes. But a fix for that is on my list. 

> I'm also having a really hard time selecting nested blocks now but can't really pinpoint it. Sometimes I hover a block, the blue borders of highlight appear, I left click, nothing happens. I left click some more, nope, dead clicks, won't select the block...

Can you elaborate a bit? Which block(s) are you testing with?

**Comment by ktmn** [comment]
> Can you elaborate a bit? Which block(s) are you testing with?

Any blocks, really. For example the same setup with columns:



No blocks are selected,
hover over the paragraph exactly where my cursor is (if you go slightly lower then the cursor will change to `text`, move up from there until the cursor is `default` again),
then left click - it won't activate the block. In 4.1.1 and earlier it did.

The intuitive expectation would be, that for what ever block the blue border is being displayed, then a left click should activate that block. And that's how it worked before already.

**Comment by ktmn** [comment]
Looks like this is the case with even the simplest blocks. 



* Red area - can't click there to activate block (as of ~4.2.0)
* Green area - can click there to activate block
* Blue area - can't even hover there (i.e. block will not appear as `.is-hovered`)

Now here's a crude example of a nested block:



As you can see the parent block barely has any green area. It would be the parent block's red area where you used to be able to activate the it, but it no longer works.

You can sometimes select the parent block by hovering in the child block's blue area though, since for some reason it won't count as child block. Maybe that's by design, but not very intuitive.

**Comment by jasmussen** [comment]
Thanks for the added context.

I'm working right now on improvements to the columns block, and will have something to show in a bit. That should fix the offset toolbars you've seen.

Regarding the clickability of individual blocks, I will look at improving that separately. I'm have a suspicion it's been like this for quite a while, as nothing has touched those margins recently. That doesn't make it not an issue, of course, and I suspect revisiting #8350 is the real solution. In the mean time, I have some ideas for an interim fix. Stay tuned.

**Comment by jasmussen** [comment]
I took a look at this today, and learned a few things from it. Take a look at this:

screenshot 2018-11-16 at 11 36 02

The coral area is the content. The red area is the hoverable/clickable area. 

We can, with seemingly no side effects, change the hoverable/clickable area to look like this:

screenshot 2018-11-16 at 11 38 03

But that doesn't benefit us a ton, because we have the sibling inserter between every block. The green area here is the hoverable area that reveals the clickable plus in the middle:

screenshot 2018-11-16 at 11 46 42

This green area incidentally enables the _hovering_ of a block, but not the clicking. And even if we expand the red clickable area, the sibling inserter blocks clicks between the blocks.

This leads us to three potential ways forward:

1: Acknowledge that the sibling inserter can't coexist with a larger click area that lets you select a block even when clicking on its margins. 
2: Reduce the _width_ of the sibling inserter hover area, so you can't hover the full width. For example we could decide you have to hover the 60% area in the middle, which would leave 20% on each side where you could click block margins to select the underlying blocks. 
3: Find a way to propogate a _click_ on the sibling inserter hoverable area, down to the block below. This is a technical challenge, and there could be side-effects.

To expand on 3, it's important to remember that blocks need to be able to exist in a variety of contexts — mobile, desktop, in various editor styles, and with various margins before and after. We know from past experieces that people _really_ love the sibling inserter, so I'm both reluctant to make its hit area smaller, and worried about changing its behavior too much since usage of it seems to have stabilized.

I'm curious of your thoughts on how best to approach this.

**Comment by ZebulanStanphill** [comment]
@jasmussen My best guess for a solution would be what I suggested back in #9202. The hover area at the bottom of the block would redirect clicks to the block it is attached to, and since it is only part of one block, it wouldn't run into issues where you have to calculate whether you clicked on the top or bottom half of it, which would be the case with the current inserter. It would also completely solve all issues of the sibling inserter overlapping content, since you would _always_ be able to access the top of a block without the sibling inserter overlapping, since your mouse has to first move into the bottom of a block for the sibling inserter to appear. I consider this to be essential in the long run, if Gutenberg intends to have an editing experience that matches the front-end.

**Comment by ktmn** [comment]
> 

@jasmussen Currently clicking on the green is a dead click. Doesn't do anything. That wasn't the case in 4.1.1. It should activate the block, unless I specifically click the + in the middle.
How come this regression happened?

Side note: the + in the middle also has a dodgy hitbox, sometimes it get's the `:active` state after click, but didn't actually do anything. But I guess that's a challenge for another day.

**Comment by jasmussen** [comment]
Interesting, I can't recall clicking through the green zone ever working. @aduth can you recall any details here?

**Comment by ktmn** [comment]
@jasmussen https://downloads.wordpress.org/plugin/gutenberg.4.1.1.zip

**Comment by aduth** [comment]
The change in behavior is likely a regression of #11018, specifically in that the event propagation stoppage which would otherwise select the block was moved from the button to the wrapping `div` (the green line in your screenshot):

https://github.com/WordPress/gutenberg/pull/11018/files#diff-2273a23d0e36880279bf677c5347dafeR62

@ktmn Would you be able to open a separate issue for this?
